### PR TITLE
Doc: Revise section on info_plist for plistlib functionality.

### PR DESCRIPTION
**Pull request by itsayellow**
See Issue #3540 
As plistlib has been added to the code to process the info_plist build argument dict, info_plist can now accept nested python objects, and python datatypes other than strings.  This update to the documentation reflects these new capabilities and includes new examples, as well as removing descriptions of limitations that no longer exist.

### CODE REVIEW

File: doc/spec-files.rst [hunk 1]
```diff
@@ -383,36 +383,47 @@ also work in retina screen::
...
-It cannot handle nested XML arrays.
-For example, if you want to modify :file:`Info.plist` to tell Mac OS X
-what filetypes your app supports, you must add a 
+|PyInstaller| creates :file:`Info.plist` from the info_plist dict
+using the Python Standard Library module plistlib_.
+plistlib can handle nested python objects (which are translated to nested
+XML.)  The following python data types are translated to the proper
+:file:`Info.plist` types: strings, integers, floats, booleans, tuples, lists,
+dictionaries (but only with string keys), Data, bytes, bytesarray, and
+datetime.datetime objects.
```
Comment (htgoebel) [inline doc/spec-files.rst]: IMO it would be okay to only refer to `plistlib` for details.

File: doc/spec-files.rst [hunk 2]
```diff
@@ -383,36 +383,47 @@ also work in retina screen::
...
-This includes the ``CFBundleExecutable`` value as well as the
-icon path and bundle identifier if you supplied them.
-Edit the ``Info.plist`` as necessary to add more items
-and save it separately.
-
-From that point on, to rebuild the app call |PyInstaller| in a shell script,
-and follow it with a statement such as::
-
-    cp -f Info.plist dist/myscript.app/Contents/Info.plist
+    app = BUNDLE(exe,
```
Comment (htgoebel) [inline doc/spec-files.rst]: I suggest having only one example, which I asumme to be enough. So the order would be something like 
- old lines up to 373 ("...parameter to the BUNDLE call")
- our new lines 390-393
- this large example
- short explanation (based on old lines 378-379 and your new lines 398-407)

File: doc/spec-files.rst [hunk 3]
```diff
@@ -383,36 +383,47 @@ also work in retina screen::
...
+             info_plist={
+             	'NSHighResolutionCapable': True
+                'CFBundleDocumentTypes': [
+                    {
+                        'CFBundleTypeName': 'My File Format'
+                        'CFBundleTypeIconFile': 'MyFileIcon.icns'
+                        'LSItemContentTypes': [
+                            'com.example.myformat'
+                            ]
+                        'LSHandlerRank': 'Owner'
```
Comment (htgoebel) [inline doc/spec-files.rst]: Trailing commas (separators) are missing at most of the lines above.

File: doc/spec-files.rst [hunk 4]
```diff
@@ -383,36 +383,47 @@ also work in retina screen::
...
+             bundle_identifier=None
+             info_plist={
+             	'NSHighResolutionCapable': True
+                'CFBundleDocumentTypes': [
+                    {
+                        'CFBundleTypeName': 'My File Format'
+                        'CFBundleTypeIconFile': 'MyFileIcon.icns'
+                        'LSItemContentTypes': [
+                            'com.example.myformat'
+                            ]
```
Comment (htgoebel) [inline doc/spec-files.rst]: Use only one line for `'LSItemContentTypes': ['com.example.myformat'],`?

File: doc/spec-files.rst [hunk 5]
```diff
@@ -383,36 +383,47 @@ also work in retina screen::
              icon=None,
              bundle_identifier=None
              info_plist={
-             	'NSHighResolutionCapable': 'True'
+             	'NSHighResolutionCapable': True
```
Comment (htgoebel) [inline doc/spec-files.rst]: Please mind the same error in line 378

**Comment by itsayellow** [comment]
Sure thing!  I appreciate good documentation and was about to volunteer if you didn't ask.  :)

For the edits, I tried to do as requested. (doh!  Sorry about the missing commas.)  Let me know how that looks.

File: doc/spec-files.rst [hunk 6]
```diff
@@ -370,49 +370,44 @@ of ``Info.plist``.)
...
-For example, when you use PyQt5,
-you can set ``NSHighResolutionCapable`` to ``True`` to let your app
-also work in retina screen::
+``info_plist=`` parameter to the BUNDLE call.  Its argument should be a
+Python dict with keys and values to be included in the :file:`Info.plist`
+file.
+|PyInstaller| creates :file:`Info.plist` from the info_plist dict
+using the Python Standard Library module plistlib_.
+plistlib can handle nested python objects (which are translated to nested
+XML.)  Python data types are translated to the proper :file:`Info.plist`
```
Comment (htgoebel) [inline doc/spec-files.rst]: "XML)." (dot behind parent)
or even
"XML), where …"

File: doc/spec-files.rst [hunk 7]
```diff
@@ -370,49 +370,44 @@ of ``Info.plist``.)
...
-and follow it with a statement such as::
-
-    cp -f Info.plist dist/myscript.app/Contents/Info.plist
+For example, when you use PyQt5,
+you can set ``NSHighResolutionCapable`` to ``True`` to let your app
+also work in retina screen.
+
+Also in the above example is the key ``CFBundleDocumentTypes``.
+This entry in :file:`Info.plist` tells Mac OS X
+what filetypes your app supports.  (see `Apple document types`_).
```
Comment (htgoebel) [inline doc/spec-files.rst]: dot and space after "supports".

File: doc/spec-files.rst [hunk 8]
```diff
@@ -370,49 +370,44 @@ of ``Info.plist``.)
...
-This includes the ``CFBundleExecutable`` value as well as the
-icon path and bundle identifier if you supplied them.
-Edit the ``Info.plist`` as necessary to add more items
-and save it separately.
-
-From that point on, to rebuild the app call |PyInstaller| in a shell script,
-and follow it with a statement such as::
-
-    cp -f Info.plist dist/myscript.app/Contents/Info.plist
+For example, when you use PyQt5,
```
Comment (htgoebel) [inline doc/spec-files.rst]: Is NSHighResolutionCapable related to PyQt5? If not, I'd remove the "when you use PyQt5,".

File: doc/spec-files.rst [hunk 9]
```diff
@@ -370,49 +370,44 @@ of ``Info.plist``.)
...
+you can set ``NSHighResolutionCapable`` to ``True`` to let your app
+also work in retina screen.
+
+Also in the above example is the key ``CFBundleDocumentTypes``.
+This entry in :file:`Info.plist` tells Mac OS X
+what filetypes your app supports.  (see `Apple document types`_).
+The value of that key in :file:`Info.plist` is a list of dicts,
+each containing up to five key:value pairs.  This is represented in the
+info_plist argument by assigning to the key ``CFBundleDocumentTypes`` a value
+that is a python list containing python dicts.
```
Comment (htgoebel) [inline doc/spec-files.rst]: I do not understand this two sentences. Maybe just remove them?

File: doc/spec-files.rst (hunk 8 above)
Comment (itsayellow) [reply doc/spec-files.rst]: This was existing in the documentation, and I didn't understand it either so I left it.  To tell you the truth I've never used NSHighResolutionCapable.  Usually the Info.plist key that's crucial for an application to have retina resolution enabled is `'NSPrincipleClass': 'NSApplication'`.

Comment (itsayellow) [reply doc/spec-files.rst]: https://developer.apple.com/library/content/documentation/General/Reference/InfoPlistKeyReference/Articles/CocoaKeys.html#//apple_ref/doc/uid/TP40009251-SW11
https://developer.apple.com/library/content/documentation/GraphicsAnimation/Conceptual/HighResolutionOSX/Explained/Explained.html

Based on those two links, I'm starting to think that NSHighResolutionCapable is a red herring and should be removed from the example.

It seems that the crucial thing to enable retina is to inform macOS that the application is a Cocoa app, using `'NSPrincipleClass': 'NSApplication'`.  Then, the default value of `NSHighResolutionCapable` should already be `True`

File: doc/spec-files.rst (hunk 9 above)
Comment (itsayellow) [reply doc/spec-files.rst]: Yes, I was trying to use what was already there, but quite right--this got confusing.  Removing them.

**Comment by itsayellow** [comment]
I've fixed the punctuation goofs (and rephrased one, let me know how it looks).

What do you think about removing `NSHighResolutionCapable` from example and discussion, and instead putting `'NSPrincipleClass': 'NSApplication'`?  (See discussion above.)

**Comment by htgoebel** [comment]
Thanks for the update. I don't like red herrings :-) 

Still I would like to have *any* other example for a boolean entry to show this has to be a boolean (and not a string or number). I would not care about using some pure example like `aBooleanValue: True`. If you know a real-life example, we could of course use it.

**Comment by itsayellow** [comment]
I agree, although my father loved (actual) red herrings...

I updated the example with a common boolean key/value.  I added 'NSPrincipleClass'.  Finally I cleaned up the explanation of the example.

Let me know how it looks!

**Comment by itsayellow** [comment]
BTW, I will be unplugged and away on vacation starting Saturday morning.  I'm happy if you want to make any changes in this pull request.

**Comment by htgoebel** [comment]
Thanks for nagging :-)

Thanks you very much for this pull-request. I appreciated the discussion and really like the result. Esp. your explanation for the boolean example is even more elaborated than I your have expected it it be.

Thanks again ans I'm looking forward for more contributions from you. Maybe you are interested in working on [other OS X related issues](https://github.com/pyinstaller/pyinstaller/issues?q=is%3Aopen+is%3Aissue+label%3A%22OS+X%22). Envoy your holidays.

**Comment by itsayellow** [comment]
Thanks for the good words, and I was happy to contribute to such a friendly project!  I will definitely take a look at the OS X issues (I already did a little bit.)

**Pull request merged**